src/main/java/org/terracotta/offheapstore/paging/UpfrontAllocatingPageSource.java

```diff
@@ -440,67 +474,164 @@ public synchronized Runnable removeAllocationThreshold(ThresholdDirection direct
-            currentChunkSize >>>= 1;
-            lastFailureAt = allocated;
-          }
-        } else {
-          buffers.add(b);
-          allocated += b.capacity();
-          
-          if (allocatorLog != null) {
-            allocatorLog.printf("%d,%d,%d,%d,%d,%d,%d,%d%n", System.nanoTime() - start, blockDuration, b.capacity(), allocated, PhysicalMemory.freePhysicalMemory(), PhysicalMemory.totalSwapSpace(), PhysicalMemory.freeSwapSpace(), PhysicalMemory.ourCommittedVirtualMemory());
-          }
-          if (allocated > nextProgressLogAt) {
-            LOGGER.info("Allocation {}% complete", (100 * allocated) / max);
-            nextProgressLogAt += progressStep;
+        if (allocatorLog != null) {
+          allocatorLog.printf("timestamp,threadid,duration,size,physfree,totalswap,freeswap,committed%n");
+        }
+
+        ExecutorService executorService = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors());
+
+        List<Future<Collection<ByteBuffer>>> futures = new ArrayList<Future<Collection<ByteBuffer>>>((int)(toAllocate / maxChunk + 1));
+
+        for (long dispatched = 0; dispatched < toAllocate; ) {
+          final int currentChunkSize = (int)Math.min(maxChunk, toAllocate - dispatched);
+          futures.add(executorService.submit(new Callable<Collection<ByteBuffer>>() {
+            @Override
+            public Collection<ByteBuffer> call() throws Exception {
+              return bufferAllocation(source, currentChunkSize, minChunk, fixed, allocatorLog, start);
+            }
+          }));
+          dispatched += currentChunkSize;
+        }
+
+        executorService.shutdown();
+
+        long allocated = 0;
+        long progressStep = Math.max(PROGRESS_LOGGING_THRESHOLD, (long)(toAllocate * PROGRESS_LOGGING_STEP_SIZE));
+        long nextProgressLogAt = progressStep;
+        AtomicBoolean wasInterrupted = new AtomicBoolean(false);
+
+        for (Future<Collection<ByteBuffer>> future : futures) {
+          Collection<ByteBuffer> result = uninterruptibleGet(future, wasInterrupted);
+          buffers.addAll(result);
+          for(ByteBuffer buffer : result) {
+            allocated += buffer.capacity();
+            if (allocated > nextProgressLogAt) {
+              LOGGER.info("Allocation {}% complete", (100 * allocated) / toAllocate);
+              nextProgressLogAt += progressStep;
+            }
           }
         }
+
+        // Set the interruption flag to tell whoever is interested
+        if(wasInterrupted.get()) {
+          Thread.currentThread().interrupt();
+        }
+
+      } finally {
+        if (allocatorLog != null) {
+          allocatorLog.close();
+        }
       }
-      if (allocatorLog != null) {
-        allocatorLog.close();
+
+      if(LOGGER.isDebugEnabled()) {
```

Comment:
This code path isn't really going to be hot enough to justify the conditionals here.